### PR TITLE
Better handling tests with global side effects

### LINKED ISSUE #9632: E2E Playwright - improve how we handle E2E tests with global side effects
With #9575 and #9577, we introduce E2E tests that create global side effects that could influence the result of other E2E tests. For this reason, those tests have to run one at a time. The flag `fullyParallel` avoid parallel execution between E2E tests in the same file, but other tests will continue to run. To avoid full parallelization, it is necessary to set the worker to 1. With [19467e7](https://github.com/woocommerce/woocommerce-blocks/pull/9577/commits/19467e7351d792eb7947f0ae42558ef2388de0e6), the worker has been set to 1 for all E2E tests, but we should avoid this approach.

### PR DESCRIPTION
**Pull request by dinhtungdu**
This PR splits our E2E tests into two types:
- Ones with side effects (tests that alter database setting or use hooks) that can only be run one at a time.
- Ones without side effects that can be run in parallel. (the main test suite).

Why? Because Playwright doesn't allow multiple type of test execution. They're [working](https://github.com/microsoft/playwright/issues/17553) on that, but haven't [released](https://github.com/microsoft/playwright/issues/19889#issuecomment-1375454753) it yet.

Two npm scripts are added for test suite with side effect:
- `npm run test:e2e:side-effects`
- `npm run test:e2e:side-effects:report`

Note: This PR doesn't enable `fullyParallel` setting for the main suite because we need to update some tests to be fully isolated. Fixing those tests will make this PR too big to review easily. So we're updating only the configuration here.

Fixes #9632 

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

#### User Facing Testing

Ensure tests are passing locally.

- `npm run test:e2e`
- `npm run test:e2e:side-effects`

* [x] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Performance Impact

This PR speeds up our Playwright tests on CI by running normal and side effects tests in parallel. Although, the performance improvement is minor because we haven't running test in parallel yet.

### Changelog

N/a.

### CODE REVIEW

File: .github/workflows/playwright.yml [hunk 1]
```diff
@@ -10,18 +10,38 @@ jobs:
...
+        env:
+          cache-name: cache-node-modules
+        with:
+          path: node_modules
+          key: ${{ runner.os }}-modified-build-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-modified-build-${{ env.cache-name }}-
+            ${{ runner.os }}-modified-build-
+            ${{ runner.os }}-modified-
+
```
Comment (imanish003) [inline .github/workflows/playwright.yml:35]: Thank you for this performance optimization. By caching the `node_modules` directory, subsequent runs of the workflow can skip the installation step if the dependencies have not changed. 🚀

File: .github/workflows/playwright.yml [hunk 2]
```diff
@@ -10,18 +10,38 @@ jobs:
...
       - name: Setup node version and npm cache
         uses: actions/setup-node@v3
         with:
           node-version-file: '.nvmrc'
           cache: 'npm'
 
       - name: Install Node dependencies
-        run: npm ci
+        if: steps.cache-node-modules.outputs.cache-hit != 'true'
+        run: npm install --no-optional --no-audit
```
Comment (imanish003) [inline .github/workflows/playwright.yml]: I'm curious, is this done primarily for performance optimization, or are there other reasons behind it as well? (I mean the reason behind not using `npm ci`)

**Review by imanish003 — COMMENTED** [review]
All the tests passed for `test:e2e`, but I'm encountering some test failures when running `npm run test:e2e:side-effects` locally. Here's the detailed report:
[playwright-report.zip](https://github.com/woocommerce/woocommerce-blocks/files/12332337/playwright-report.zip)

**Review by imanish003 — APPROVED** [review]
Hey Tung! I've re-run the tests, and I'm delighted to share that all the E2E tests are passing smoothly now. 🙌🏻
Thank you so much for your efforts on this. I'm hopeful that we'll soon be able to enable `fullyParallel` for the main suite. Keep up the fantastic work! 🚀

PS: I did leave one comment in the previous review. Although, I don't think that's a blocker, therefore pre-approving the PR 🚢

File: .github/workflows/playwright.yml (hunk 2 above)
Comment (dinhtungdu) [reply .github/workflows/playwright.yml]: Good point, let me change this back to `npm ci`

**Review by thealexandrelara — APPROVED** [review]
I just tested and all tests are passing ✅ 

I believe we should enhance our documentation regarding the e2e tests to include this new npm script. As we're now splitting the e2e tests into multiple scripts, it would be good if we can update the README for [WooCommerce Blocks Playwright End to End Tests](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tests/e2e/README.md). However, if you think there's another location that would also benefit from this update, please don't hesitate to make the necessary changes.

I'm pre-approving this PR since everything is working as expected. Great work! 🚀

**Comment by dinhtungdu** [comment]
@thealexandrelara really nice catch! I https://github.com/woocommerce/woocommerce-blocks/pull/10508/commits/c1cf8bd097807a9ba0f26a93b59240cc37ed9c27 the documentation to mention the new command. Thank you so much!

**Pull request merged**